### PR TITLE
QUIC support for wolfSSL

### PR DESCRIPTION
**Pull request by icing**
# Description

This PR adds QUIC support to wolfSSL. 

Please see the supplied QUIC.md for a complete description.

# Testing

- test cases have been added in `tests/quic.c` and run as part of `make test`.
- this PR has been verified in the [ngtcp2](https://github.com/ngtcp2/ngtcp2) project to achieve inter-op with itself, openssl and other public QUIC implementations.

# Checklist

 - [x] added tests
 - [ ] updated/added doxygen
tbd. when changes and new API has been reviewed.
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation
see QUIC.md as a base for further documentation updates

### CODE REVIEW

**Comment by dgarske** [comment]
OK to test

**Comment by icing** [comment]
I have no access to the 2 last failing test suites. So, if someone could figure out what is needed to have them pass, let me know.

**Comment by dgarske** [comment]
> I have no access to the 2 last failing test suites. So, if someone could figure out what is needed to have them pass, let me know.

```
./configure --enable-sniffer && make check
FAIL: scripts/sniffer-testsuite.test
```

```
-----------------------------------------------
Detecting new files that have been added
-----------------------------------------------
added: QUIC.md
added: src/quic.c
added: tests/quic.c
added: wolfssl/quic.h
Total new files to check: 4
-----------------------------------------------
Checking if files are included in an include.am
-----------------------------------------------
./src/include.am:src_libwolfssl_la_SOURCES += src/quic.c
./tests/include.am:                  tests/quic.c \
./wolfssl/include.am:                         wolfssl/quic.h \
AUTOCONF NEW FILE DETECTOR FAILED!
```

**Comment by icing** [comment]
> ```
> ./configure --enable-sniffer && make check
> FAIL: scripts/sniffer-testsuite.test
> ```

Thank you! The latest addition fixes that for me. Let's see what CI says.

Also as a note, if I configure

```
> ./configure --enable-all --enable-sniffer && make check
```

I get failures in ocsp-stapling and unit.tests, but after some research I found that those also happen in master. Just FYI.

File: src/QUIC.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
+# QUIC in wolfSSL
```
Comment (dgarske) [inline src/QUIC.md:1]: Please move this mark down file. Either in docs or root.

File: tests/quic.c [hunk 2]
```diff
@@ -0,0 +1,1390 @@
...
+    /* Should have received a session ticket, save the session */
+    AssertTrue(tclient.ticket_len > 0);
+    AssertNotNull(session = wolfSSL_get1_session(tclient.ssl));
+    QuicTestContext_free(&tclient);
+    QuicTestContext_free(&tserver);
+
+    /* QUIC requires 0 or 0xffffffff as only allowed values.
+     * Since we enabled early data in the server that created the session,
+     * we need to see it here. */
+    unsigned int max_early_sz = wolfSSL_SESSION_get_max_early_data(session);
```
Comment (dgarske) [inline tests/quic.c:1228]: Please have all variable declarations at top of function or brace:
```tests/quic.c:1228:5: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
 1228 |     unsigned int max_early_sz = wolfSSL_SESSION_get_max_early_data(session);
      |     ^~~~~~~~
```

**Review by dgarske — CHANGES_REQUESTED** [review]
Some minor cast warnings:

```
./configure CC=g++ --enable-quic

In file included from tests/quic.c:29:

tests/quic.c: In function ‘int test_quic_crypt()’:
./wolfssl/wolfcrypt/types.h:508:67: error: invalid conversion from ‘void*’ to ‘uint8_t*’ {aka ‘unsigned char*’} [-fpermissive]
  508 |                 #define XMALLOC(s, h, t)     ((void)(h), (void)(t), wolfSSL_Malloc((s)))
      |                                              ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
      |                                                                   |
      |                                                                   void*
./tests/unit.h:52:49: note: in definition of macro ‘Assert’

   52 | #define Assert(test, description, result) if (!(test)) Fail(description, result)
      |                                                 ^~~~

tests/quic.c:353:9: note: in expansion of macro ‘AssertNotNull’

  353 |         AssertNotNull(encrypted = XMALLOC(enc_len, NULL, DYNAMIC_TYPE_TMP_BUFFER));
      |         ^~~~~~~~~~~~~

tests/quic.c:353:35: note: in expansion of macro ‘XMALLOC’

  353 |         AssertNotNull(encrypted = XMALLOC(enc_len, NULL, DYNAMIC_TYPE_TMP_BUFFER));
      |                                   ^~~~~~~
./wolfssl/wolfcrypt/types.h:508:67: error: invalid conversion from ‘void*’ to ‘uint8_t*’ {aka ‘unsigned char*’} [-fpermissive]
  508 |                 #define XMALLOC(s, h, t)     ((void)(h), (void)(t), wolfSSL_Malloc((s)))
      |                                              ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
      |                                                                   |
      |                                                                   void*
./tests/unit.h:52:49: note: in definition of macro ‘Assert’

   52 | #define Assert(test, description, result) if (!(test)) Fail(description, result)
      |                                                 ^~~~

tests/quic.c:354:9: note: in expansion of macro ‘AssertNotNull’

  354 |         AssertNotNull(decrypted = XMALLOC(dec_len, NULL, DYNAMIC_TYPE_TMP_BUFFER));
      |         ^~~~~~~~~~~~~

tests/quic.c:354:35: note: in expansion of macro ‘XMALLOC’

  354 |         AssertNotNull(decrypted = XMALLOC(dec_len, NULL, DYNAMIC_TYPE_TMP_BUFFER));
      |                                   ^~~~~~~

tests/quic.c: In function ‘void QuicTestContext_init(QuicTestContext*, WOLFSSL_CTX*, const char*, int)’:

tests/quic.c:432:66: error: narrowing conversion of ‘-1’ from ‘int’ to ‘byte’ {aka ‘unsigned char’} [-Wnarrowing]
  432 |     static const byte tp_params_s[] = {7, 6, 5, 4, 3, 2, 1, 0, -1};
      |                                                                  ^

tests/quic.c: In function ‘int ctx_set_encryption_secrets(WOLFSSL*, WOLFSSL_ENCRYPTION_LEVEL, const uint8_t*, const uint8_t*, size_t)’:

tests/quic.c:476:48: error: invalid conversion from ‘void*’ to ‘QuicTestContext*’ [-fpermissive]
  476 |     QuicTestContext *ctx = wolfSSL_get_app_data(ssl);
      |                            ~~~~~~~~~~~~~~~~~~~~^~~~~
      |                                                |
      |                                                void*

tests/quic.c: In function ‘int ctx_add_handshake_data(WOLFSSL*, WOLFSSL_ENCRYPTION_LEVEL, const uint8_t*, size_t)’:

tests/quic.c:495:48: error: invalid conversion from ‘void*’ to ‘QuicTestContext*’ [-fpermissive]
  495 |     QuicTestContext *ctx = wolfSSL_get_app_data(ssl);
      |                            ~~~~~~~~~~~~~~~~~~~~^~~~~
      |                                                |
      |                                                void*

tests/quic.c:505:31: error: invalid conversion from ‘void*’ to ‘OutputBuffer*’ [-fpermissive]
  505 |             out->next = calloc(1, sizeof(OutputBuffer));
      |                         ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
      |                               |
      |                               void*

tests/quic.c: In function ‘int ctx_flush_flight(WOLFSSL*)’:

tests/quic.c:525:48: error: invalid conversion from ‘void*’ to ‘QuicTestContext*’ [-fpermissive]
  525 |     QuicTestContext *ctx = wolfSSL_get_app_data(ssl);
      |                            ~~~~~~~~~~~~~~~~~~~~^~~~~
      |                                                |
      |                                                void*

tests/quic.c: In function ‘int ctx_send_alert(WOLFSSL*, WOLFSSL_ENCRYPTION_LEVEL, uint8_t)’:

tests/quic.c:534:48: error: invalid conversion from ‘void*’ to ‘QuicTestContext*’ [-fpermissive]
  534 |     QuicTestContext *ctx = wolfSSL_get_app_data(ssl);
      |                            ~~~~~~~~~~~~~~~~~~~~^~~~~
      |                                                |
      |                                                void*

make[2]: *** [Makefile:6497: tests/unit_test-quic.o] Error 1

make[2]: *** Waiting for unfinished jobs....

In file included from ./wolfssl/wolfcrypt/misc.h:32,
                 from ./wolfcrypt/src/misc.c:37,
                 from src/quic.c:34:

src/quic.c: In function ‘QuicRecord* quic_record_make(WOLFSSL*, WOLFSSL_ENCRYPTION_LEVEL, const uint8_t*, size_t)’:
./wolfssl/wolfcrypt/types.h:508:67: error: invalid conversion from ‘void*’ to ‘QuicRecord*’ [-fpermissive]
  508 |                 #define XMALLOC(s, h, t)     ((void)(h), (void)(t), wolfSSL_Malloc((s)))
      |                                              ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
      |                                                                   |
      |                                                                   void*

src/quic.c:76:10: note: in expansion of macro ‘XMALLOC’

   76 |     qr = XMALLOC(sizeof(*qr), ssl->heap, DYNAMIC_TYPE_TMP_BUFFER);
      |          ^~~~~~~
./wolfssl/wolfcrypt/types.h:508:67: error: invalid conversion from ‘void*’ to ‘uint8_t*’ {aka ‘unsigned char*’} [-fpermissive]
  508 |                 #define XMALLOC(s, h, t)     ((void)(h), (void)(t), wolfSSL_Malloc((s)))
      |                                              ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
      |                                                                   |
      |                                                                   void*

src/quic.c:89:20: note: in expansion of macro ‘XMALLOC’

   89 |         qr->data = XMALLOC(qr->capacity, ssl->heap, DYNAMIC_TYPE_TMP_BUFFER);
      |                    ^~~~~~~

src/quic.c: In function ‘int quic_record_append(WOLFSSL*, QuicRecord*, const uint8_t*, size_t, size_t*)’:
./wolfssl/wolfcrypt/types.h:514:61: error: invalid conversion from ‘void*’ to ‘uint8_t*’ {aka ‘unsigned char*’} [-fpermissive]
  514 |                 #define XREALLOC(p, n, h, t) wolfSSL_Realloc((p), (n))
      |                                              ~~~~~~~~~~~~~~~^~~~~~~~~~
      |                                                             |
      |                                                             void*

src/quic.c:131:30: note: in expansion of macro ‘XREALLOC’

  131 |             uint8_t *ndata = XREALLOC(qr->data, qr->len, ssl->head, DYNAMIC_TYPE_TMP_BUFFER);
      |                              ^~~~~~~~

src/quic.c: In function ‘word32 quic_record_transfer(QuicRecord*, byte*, word32)’:

src/quic.c:186:44: error: enumeral and non-enumeral type in conditional expression [-Werror=extra]
  186 |         rlen = (qr->len <= MAX_RECORD_SIZE)? qr->len : MAX_RECORD_SIZE;
      |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~

In file included from ./wolfssl/wolfcrypt/misc.h:32,
                 from ./wolfcrypt/src/misc.c:37,
                 from src/quic.c:34:

src/quic.c: In function ‘const QuicTransportParam* QuicTransportParam_new(const uint8_t*, size_t, void*)’:
./wolfssl/wolfcrypt/types.h:508:67: error: invalid conversion from ‘void*’ to ‘QuicTransportParam*’ [-fpermissive]
  508 |                 #define XMALLOC(s, h, t)     ((void)(h), (void)(t), wolfSSL_Malloc((s)))
      |                                              ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
      |                                                                   |
      |                                                                   void*

src/quic.c:212:10: note: in expansion of macro ‘XMALLOC’

  212 |     tp = XMALLOC(sizeof(*tp), heap, DYNAMIC_TYPE_TLSX);
      |          ^~~~~~~
./wolfssl/wolfcrypt/types.h:508:67: error: invalid conversion from ‘void*’ to ‘const uint8_t*’ {aka ‘const unsigned char*’} [-fpermissive]
  508 |                 #define XMALLOC(s, h, t)     ((void)(h), (void)(t), wolfSSL_Malloc((s)))
      |                                              ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
      |                                                                   |
      |                                                                   void*

src/quic.c:214:16: note: in expansion of macro ‘XMALLOC’

  214 |     tp->data = XMALLOC(len, heap, DYNAMIC_TYPE_TLSX);
      |                ^~~~~~~

src/quic.c: In function ‘const QuicTransportParam* QuicTransportParam_dup(const QuicTransportParam*, void*)’:
./wolfssl/wolfcrypt/types.h:508:67: error: invalid conversion from ‘void*’ to ‘QuicTransportParam*’ [-fpermissive]
  508 |                 #define XMALLOC(s, h, t)     ((void)(h), (void)(t), wolfSSL_Malloc((s)))
      |                                              ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
      |                                                                   |
      |                                                                   void*

src/quic.c:227:11: note: in expansion of macro ‘XMALLOC’

  227 |     tp2 = XMALLOC(sizeof(*tp2), heap, DYNAMIC_TYPE_TLSX);
      |           ^~~~~~~
./wolfssl/wolfcrypt/types.h:508:67: error: invalid conversion from ‘void*’ to ‘const uint8_t*’ {aka ‘const unsigned char*’} [-fpermissive]
  508 |                 #define XMALLOC(s, h, t)     ((void)(h), (void)(t), wolfSSL_Malloc((s)))
      |                                              ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
      |                                                                   |
      |                                                                   void*

src/quic.c:229:17: note: in expansion of macro ‘XMALLOC’

  229 |     tp2->data = XMALLOC(tp->len, heap, DYNAMIC_TYPE_TLSX);
      |                 ^~~~~~~

src/quic.c: In function ‘int wolfSSL_quic_send_internal(WOLFSSL*)’:

src/quic.c:740:27: error: invalid conversion from ‘int’ to ‘WOLFSSL_ENCRYPTION_LEVEL’ {aka ‘wolfssl_encryption_level_t’} [-fpermissive]
  740 |                 ssl->quic.output_rec_level, (const uint8_t*)output, len);
      |                 ~~~~~~~~~~^~~~~~~~~~~~~~~~
      |                           |
      |                           int
```

File: src/quic.c [hunk 3]
```diff
@@ -0,0 +1,1225 @@
+/* quic.c
+ *
+ * Copyright (C) 2006-2021 wolfSSL Inc.
```
Comment (SparkiDev) [inline src/quic.c:3]: 2022

File: src/tls.c [hunk 4]
```diff
@@ -10266,6 +10269,90 @@ int TLSX_EarlyData_Use(WOLFSSL* ssl, word32 maxSz)
...
+/******************************************************************************/
+/* QUIC transport parameter extension                                         */
+/******************************************************************************/
+#ifdef WOLFSSL_QUIC
+
+static word16 TLSX_QuicTP_GetSize(TLSX* extension)
+{
+    const QuicTransportParam *tp = extension->data;
+
+    return tp? tp->len : 0;
```
Comment (SparkiDev) [inline src/tls.c:10281]: Space before '?'

File: src/tls.c [hunk 5]
```diff
@@ -10266,6 +10269,90 @@ int TLSX_EarlyData_Use(WOLFSSL* ssl, word32 maxSz)
...
 #endif
 
+/******************************************************************************/
+/* QUIC transport parameter extension                                         */
+/******************************************************************************/
+#ifdef WOLFSSL_QUIC
+
+static word16 TLSX_QuicTP_GetSize(TLSX* extension)
+{
+    const QuicTransportParam *tp = extension->data;
```
Comment (SparkiDev) [inline src/tls.c:10279]: '*' after type, not before variable name.

File: src/tls.c [hunk 6]
```diff
@@ -10408,6 +10495,14 @@ void TLSX_FreeAll(TLSX* list, void* heap)
                 break;
 #endif
 
+    #ifdef WOLFSSL_QUIC
+            case TLSX_KEY_QUIC_TP_PARAMS:
+                FALL_THROUGH;
+            case TLSX_KEY_QUIC_TP_PARAMS_DRAFT:
+                QuicTransportParam_free((QuicTransportParam*)extension->data, heap);
```
Comment (SparkiDev) [inline src/tls.c:10502]: For consistency, add a define like QTP_FREE that maps to something.

File: src/tls.c [hunk 7]
```diff
@@ -10750,6 +10853,15 @@ static int TLSX_Write(TLSX* list, byte* output, byte* semaphore,
             case TLSX_USE_SRTP:
                 offset += SRTP_WRITE((TlsxSrtp*)extension->data, output+offset);
                 break;
+#endif
+#ifdef WOLFSSL_QUIC
+            case TLSX_KEY_QUIC_TP_PARAMS:
+                FALL_THROUGH;
+            case TLSX_KEY_QUIC_TP_PARAMS_DRAFT:
+                WOLFSSL_MSG("QUIC transport parameter to write");
+                offset += TLSX_QuicTP_Write((QuicTransportParam*)extension->data,
```
Comment (SparkiDev) [inline src/tls.c:10862]: For consistency, use a define like QTP_WRITE.

File: src/tls.c [hunk 8]
```diff
@@ -12384,6 +12498,34 @@ int TLSX_Parse(WOLFSSL* ssl, const byte* input, word16 length, byte msgType,
...
+                        msgType != server_hello &&
+                        msgType != encrypted_extensions) {
+                    return EXT_NOT_ALLOWED;
+                }
+                else if (!IsAtLeastTLSv1_3(ssl->version) &&
+                         msgType == encrypted_extensions) {
+                    return EXT_NOT_ALLOWED;
+                }
+                else if (WOLFSSL_IS_QUIC(ssl)) {
+                    ret = TLSX_QuicTP_Parse(ssl, input + offset, size, type, msgType);
```
Comment (SparkiDev) [inline src/tls.c:12521]: For consistency, use define like QTP_PARSE.

File: configure.ac [hunk 9]
```diff
@@ -1031,6 +1032,25 @@ then
...
+
+if test "$ENABLED_QUIC" = "yes"
+then
+    if test "x$ENABLED_TLS13" = "xno"
+    then
+        AC_MSG_ERROR([TLS 1.3 is disabled - necessary for QUIC])
+    fi
+    AM_CFLAGS="$AM_CFLAGS -DWOLFSSL_QUIC"
+    # QUIC proto handlers need app_data at WOLFSSL*
+    AM_CFLAGS="$AM_CFLAGS -DHAVE_EX_DATA"
```
[the remark this answers is not in the document]
Comment (icing) [reply configure.ac:1050]: The user in this case is the QUIC protocol implementation that installs the callbacks. This is the design decision made by the quictls/openssl fork and followed in boringssl implementation. gnutls uses its equivalent `gnutls_session_get_ptr()`.

We could of course add an extra `void *quic_ctx` to set/get in wolfSSL, but that seems superfluous.

File: src/tls13.c [hunk 10]
```diff
@@ -3528,13 +3564,23 @@ int SendTls13ClientHello(WOLFSSL* ssl)
...
     else {
     #ifdef WOLFSSL_TLS13_MIDDLEBOX_COMPAT
-        args->output[args->idx++] = ID_LEN;
-        XMEMCPY(args->output + args->idx, ssl->arrays->clientRandom, ID_LEN);
-        args->idx += ID_LEN;
-    #else
-        /* TLS v1.3 does not use session id - 0 length. */
-        args->output[args->idx++] = 0;
+        if (ssl->options.tls13MiddleBoxCompat) {
+            if (WOLFSSL_IS_QUIC(ssl)) {
```
Comment (SparkiDev) [inline src/tls13.c:3568]: Can we get here?

File: src/QUIC.md (hunk 1 above)
Comment (icing) [reply src/QUIC.md:1]: Will move to doc.

**Review by SparkiDev — CHANGES_REQUESTED** [review]
80 characters per line thanks. Noticeably quic.c.

We are using lots of OpenSSL type in new public APIs.
Is this to fit in with other products?

File: src/tls.c (hunk 5 above)
Comment (icing) [reply src/tls.c:10279]: This hurts, but I see now that all the code follows this. Will change.

Comment (SparkiDev) [reply src/tls.c:10279]: Did me too!
Leave quic.c alone.

File: src/tls13.c (hunk 10 above)
Comment (icing) [reply src/tls13.c:3568]: Not any more. Will remove.

File: src/tls.c (hunk 8 above)
Comment (icing) [reply src/tls.c]: Added.

File: src/tls.c (hunk 7 above)
Comment (icing) [reply src/tls.c]: Added.

File: src/tls.c (hunk 6 above)
Comment (icing) [reply src/tls.c]: Added.

File: src/tls.c (hunk 5 above)
Comment (icing) [reply src/tls.c]: Too late, already done. ;)

File: src/tls.c (hunk 4 above)
Comment (icing) [reply src/tls.c]: Done.

File: src/quic.c (hunk 3 above)
Comment (icing) [reply src/quic.c]: Done.

File: tests/quic.c (hunk 2 above)
Comment (icing) [reply tests/quic.c]: Changed.

**Comment by icing** [comment]
> Some minor cast warnings:

Should all be resolved now.

**Comment by icing** [comment]
> 80 characters per line thanks. Noticeably quic.c.

Done.

> We are using lots of OpenSSL type in new public APIs. Is this to fit in with other products?

Which ones specifically? I think I used the WOLFSSL things where possible. Are your talking about the use of `uint8_t`? That I did to be more compatible to the quictls/openssl fork, yes. It's not exactly required and we could change.

**Comment by SparkiDev** [comment]
APIs like:
  wolfSSL_quic_aead_is_gcm
  wolfSSL_quic_get_md
  wolfSSL_quic_get_hp
  wolfSSL_quic_aead_encrypt
  wolfSSL_quic_hkdf_expand

Use OpenSSL types.

**Comment by icing** [comment]
> APIs like: wolfSSL_quic_aead_is_gcm wolfSSL_quic_get_md wolfSSL_quic_get_hp wolfSSL_quic_aead_encrypt wolfSSL_quic_hkdf_expand
> 
> Use OpenSSL types.

Ah, I see now what you mean. For me, anything with prefix `WOLFSSL_` was a wolfSSL type and not an openssl one. But WOLFSSL_EVP_CIPHER is from `wolfssl/openssl/evp.h`, so for you it is an openssl type. Got you.

The motivation for this was that I tried to stay close to the quictls/openssl patch *and* that crypto internals are not my strength. It was not immediately obvious to me how the _EVP_ functions could be replaced with something like `wc_Tls13_HKDF_Expand_Label()`without deep diving into the crypt layer.

If you think there is value in avoiding the _EVP_ part of wolfSSL, I'd need either some guidance or someone with deeper understanding needs to write a PR on top of this one with the changes. I can't tell what the best way to approach this would be.

**Comment by SparkiDev** [comment]
That's fine we'll add different APIs later if needed.
Thanks!

**Review by SparkiDev — APPROVED** [review]

**Review by dgarske — APPROVED** [review]

**Comment by dgarske** [comment]
@darktohka , please resolve minor merge conflict.

**Review by dgarske — CHANGES_REQUESTED** [review]
Spelling:

```
diff --git a/src/quic.c b/src/quic.c
index 79dae7371..91f7f257a 100644
--- a/src/quic.c
+++ b/src/quic.c
@@ -451,7 +451,7 @@ int wolfSSL_quic_add_transport_extensions(WOLFSSL* ssl, int msg_type)
     }

     if (is_resp) {
-        /* server repsonse: time to decide which version to use */
+        /* server response: time to decide which version to use */
         if (ssl->quic.transport_peer && ssl->quic.transport_peer_draft) {
             if (ssl->quic.transport_version == TLSX_KEY_QUIC_TP_PARAMS_DRAFT) {
                 ret = TLSX_QuicTP_Use(ssl,
@@ -536,7 +536,7 @@ void wolfSSL_set_quic_early_data_enabled(WOLFSSL* ssl, int enabled)
      * not started yet.
      * This function is part of the quictls/openssl API and does
      * not return any error, sadly. So we just ignore any
-     * unsuccessfull use. But we can produce some warnings.
+     * unsuccessful use. But we can produce some warnings.
      */
     if (!WOLFSSL_IS_QUIC(ssl)) {
         WOLFSSL_MSG("wolfSSL_set_quic_early_data_enabled: not a QUIC SSL");
@@ -754,7 +754,7 @@ static int wolfSSL_quic_send_internal(WOLFSSL* ssl)
             aret = ssl->quic.method->add_handshake_data(ssl,
                 ssl->quic.output_rec_level, (const uint8_t*)output, len);
             if (aret != 1) {
-                /* The application has an error. General desaster. */
+                /* The application has an error. General disaster. */
                 WOLFSSL_MSG("WOLFSSL_QUIC_SEND application failed");
                 ret = FWRITE_ERROR;
                 goto cleanup;
```

CC=g++ build errors:

```
src/quic.c:189:45: error: enumeral and non-enumeral type in conditional expression [-Werror=extra]
  189 |         rlen = (qr->len <= MAX_RECORD_SIZE) ? qr->len : MAX_RECORD_SIZE;
      |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
tests/quic.c: In function ‘void QuicTestContext_init(QuicTestContext*, WOLFSSL_CTX*, const char*, int)’:
tests/quic.c:434:66: error: narrowing conversion of ‘-1’ from ‘int’ to ‘byte’ {aka ‘unsigned char’} [-Wnarrowing]
  434 |     static const byte tp_params_s[] = {7, 6, 5, 4, 3, 2, 1, 0, -1};
      |                                                                  ^
tests/quic.c: In function ‘int new_session_cb(WOLFSSL*, WOLFSSL_SESSION*)’:
tests/quic.c:1277:48: error: invalid conversion from ‘void*’ to ‘QuicTestContext*’ [-fpermissive]
 1277 |     QuicTestContext *ctx = wolfSSL_get_app_data(ssl);
      |                            ~~~~~~~~~~~~~~~~~~~~^~~~~
      |                                                |
      |                                                void*
```

**Comment by icing** [comment]
Should be fixed now. Thanks for reviewing.

**Review by dgarske — APPROVED** [review]
This all looks great!

Can you please squash merge these commits?

```
git checkout quic
git rebase master
git checkout master
git merge --squash quic
git branch -m quic quic_old
git checkout -b quic
git commit
git push -f
```

**Comment by bagder** [comment]
I am not in a position to review this wolfSSL code, but I'm eager to try this out as soon as this has merged - as part of the greater work by @icing, as I hope to use wolfSSL with ngtcp2 to do full fledged HTTP/3 in curl with this stack.

**Pull request merged**